### PR TITLE
Refactor the CLI.

### PR DESCRIPTION
**Pull request by trivialfis**
* Enable parameter validation.
* Enable JSON.
* Catch `dmlc::Error`.
* Show help message.

### CODE REVIEW

**Comment by trivialfis** [comment]
The help message obtained by running `xgboost` or `xgboost -h`:

```
Usage: xgboost [ -h ] [ config file ] [ arguments ]

  Options and arguments:
    -h, --help
       Print this message.

    arguments
       Extra parameters that are not specified in config file, see below.

  Config file specifies the configuration for both training and testing.  Each line
  containing the [attribute] = [value] configuration.

  General XGBoost parameters:

    https://xgboost.readthedocs.io/en/latest/parameter.html"

  Command line interface specfic parameters:

    task : {'dump', 'pred', 'train'},optional, default='train'
        Task to be performed by the CLI program.
    eval_train : boolean, optional, default=0
        Whether evaluate on training data during training.
    num_round : int, optional, default='10'
        Number of boosting iterations
    save_period : int, optional, default='0'
        The period to save the model, 0 means only save final model.
    train_path : string, optional, default='NULL'
        Training data path.
    test_path : string, optional, default='NULL'
        Test data path.
    model_in : string, optional, default='NULL'
        Input model path, if any.
    model_out : string, optional, default='NULL'
        Output model path, if any.
    model_dir : string, optional, default='./'
        Output directory of period checkpoint.
    name_pred : string, optional, default='pred.txt'
        Name of the prediction file.
    dsplit : {'auto', 'col', 'row'},optional, default='auto'
        Data split mode.
    ntree_limit : int, optional, default='0'
        Number of trees used for prediction, 0 means use all trees.
    pred_margin : boolean, optional, default=0
        Whether to predict margin value instead of probability.
    dump_stats : int, optional, default='0'
        Whether dump the model statistics.
    dump_format : string, optional, default='text'
        What format to dump the model in.
    name_fmap : string, optional, default='NULL'
        Name of the feature map file.
    name_dump : string, optional, default='dump.txt'
        Name of the output dump text file.
    eval[NAME]: With NAME as the name of evaluation data.

  Example:  train.conf

    booster = gbtree
    objective = reg:squarederror
    eta = 1.0
    gamma = 1.0
    seed = 0
    min_child_weight = 0
    max_depth = 3

    num_round = 2
    save_period = 0
    data = "demo/data/agaricus.txt.train?format=libsvm"
    eval[test] = "demo/data/agaricus.txt.test?format=libsvm"

  See demo/ directory in XGBoost for more examples.

```

**Comment by hcho3** [comment]
@trivialfis Should this be part of 1.1.0 release?

**Comment by trivialfis** [comment]
@hcho3 Not needed.

File: src/cli_main.cc [hunk 1]
```diff
@@ -145,221 +151,354 @@ struct CLIParam : public XGBoostParameter<CLIParam> {
...
-    std::unique_ptr<dmlc::Stream> fs(
-        dmlc::Stream::Create(param.name_fmap.c_str(), "r"));
-    dmlc::istream is(fs.get());
-    fmap.LoadText(is);
+  Config file specifies the configuration for both training and testing.  Each line
+  containing the [attribute] = [value] configuration.
+
+  General XGBoost parameters:
+
+    https://xgboost.readthedocs.io/en/latest/parameter.html"
```
Comment (trivialfis) [inline src/cli_main.cc:390]: Remove the trailing ".

**Comment by trivialfis** [comment]
ping @hcho3 @RAMitchell . ;-)  The important part about this PR is, it catches all `dmlc:Error` like our C API, so users don't get segfault for user error.

File: src/cli_main.cc [hunk 2]
```diff
@@ -145,221 +151,376 @@ struct CLIParam : public XGBoostParameter<CLIParam> {
...
+class CLI {
+  CLIParam param_;
+  std::unique_ptr<Learner> learner_;
+  enum Print {
+    kNone,
+    kVersion,
+    kHelp
+  } print_info_ {kNone};
+
+  int ResetLearner(std::vector<std::shared_ptr<DMatrix>> const& matrics) {
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/cli_main.cc:178]: Yup, good catch!

**Pull request merged**